File: backend/src/fhir/fhir-client.js
const axios = require('axios');
const { decrypt } = require('../auth/ihce-oauth');

class FhirClient {
  constructor(baseUrl) {
    this.baseUrl = baseUrl || process.env.IHCE_FHIR_BASE_URL;
  }

  async getPatient(id, encryptedToken) {
    const token = decrypt(encryptedToken);
    try {
      const response = await axios.get(`${this.baseUrl}/Patient/${id}`, {
        headers: {
          'Authorization': `Bearer ${token}`,
          'Accept': 'application/fhir+json'
        }
      });
      return response.data;
    } catch (error) {
      console.error(`Error fetching Patient ${id}:`, error.response?.data || error.message);
      throw error;
    }
  }

  async getRDA(patientId, encryptedToken) {
    const token = decrypt(encryptedToken);
    try {
      // Composition with type RDA (Resumen Digital de Atención)
      // Reference: https://co.fhir.guide/core/
      const response = await axios.get(`${this.baseUrl}/Composition`, {
        params: {
          patient: patientId,
          type: 'http://loinc.org|60591-5' // Example LOINC for Patient summary
        },
        headers: {
          'Authorization': `Bearer ${token}`,
          'Accept': 'application/fhir+json'
        }
      });
      return response.data;
    } catch (error) {
      console.error(`Error fetching RDA for Patient ${patientId}:`, error.response?.data || error.message);
      throw error;
    }
  }

  async searchPatient(params, encryptedToken) {
    const token = decrypt(encryptedToken);
    try {
      const response = await axios.get(`${this.baseUrl}/Patient`, {
        params: params,
        headers: {
          'Authorization': `Bearer ${token}`,
          'Accept': 'application/fhir+json'
        }
      });
      return response.data;
    } catch (error) {
      console.error('Error searching patient:', error.response?.data || error.message);
      throw error;
    }
  }
}

module.exports = { FhirClient };

File: lib/features/sync/data/fhir_client.dart
// TODO: migrate from data/ — moved to infrastructure/services/fhir_client.dart
// This file is kept temporarily. Remove after verifying new imports work.

/// FHIR Client for IHCE Colombia
/// Native Dart implementation - no backend needed
/// Uses flutter_appauth for OAuth and http for FHIR API calls
library;

import 'dart:convert';
import 'package:http/http.dart' as http;

class FhirClient {
  final String baseUrl;
  final http.Client _client;

  FhirClient({
    String? baseUrl,
    http.Client? client,
  })  : baseUrl = baseUrl ?? 'https://fhir.ihcecol.gov.co/fhir',
        _client = client ?? http.Client();

  /// Fetch a Patient resource by ID
  Future<Map<String, dynamic>> getPatient(
    String id,
    String accessToken,
  ) async {
    try {
      final response = await _client.get(
        Uri.parse('$baseUrl/Patient/$id'),
        headers: {
          'Authorization': 'Bearer $accessToken',
          'Accept': 'application/fhir+json',
        },
      );

      if (response.statusCode == 200) {
        return jsonDecode(response.body) as Map<String, dynamic>;
      } else {
        throw FhirException(
          'Error fetching Patient $id',
          response.statusCode,
          response.body,
        );
      }
    } catch (e) {
      if (e is FhirException) rethrow;
      throw FhirException('Network error fetching Patient: $e', 0, null);
    }
  }

  /// Fetch RDA (Resumen Digital de Atención) for a patient
  /// Uses Composition resource with LOINC code for patient summary
  Future<Map<String, dynamic>> getRDA(
    String patientId,
    String accessToken,
  ) async {
    try {
      final response = await _client.get(
        Uri.parse('$baseUrl/Composition').replace(
          queryParameters: {
            'patient': patientId,
            'type': 'http://loinc.org|60591-5', // Patient Summary
            '_sort': '-date',
            '_count': '1',
          },
        ),
        headers: {
          'Authorization': 'Bearer $accessToken',
          'Accept': 'application/fhir+json',
        },
      );

      if (response.statusCode == 200) {
        return jsonDecode(response.body) as Map<String, dynamic>;
      } else {
        throw FhirException(
          'Error fetching RDA for Patient $patientId',
          response.statusCode,
          response.body,
        );
      }
    } catch (e) {
      if (e is FhirException) rethrow;
      throw FhirException('Network error fetching RDA: $e', 0, null);
    }
  }

  /// Search Patient by identifier (document number)
  Future<Map<String, dynamic>> searchPatient({
    String? identifier,
    String? name,
    String? birthDate,
    required String accessToken,
  }) async {
    final params = <String, String>{};
    if (identifier != null) params['identifier'] = identifier;
    if (name != null) params['name'] = name;
    if (birthDate != null) params['birthdate'] = birthDate;

    if (params.isEmpty) {
      throw ArgumentError('At least one search parameter required');
    }

    try {
      final response = await _client.get(
        Uri.parse('$baseUrl/Patient').replace(queryParameters: params),
        headers: {
          'Authorization': 'Bearer $accessToken',
          'Accept': 'application/fhir+json',
        },
      );

      if (response.statusCode == 200) {
        return jsonDecode(response.body) as Map<String, dynamic>;
      } else {
        throw FhirException(
          'Error searching Patient',
          response.statusCode,
          response.body,
        );
      }
    } catch (e) {
      if (e is FhirException) rethrow;
      throw FhirException('Network error searching Patient: $e', 0, null);
    }
  }

  /// Fetch specific resource type by reference
  Future<Map<String, dynamic>?> fetchResource(
    String resourceType,
    String id,
    String accessToken,
  ) async {
    try {
      final response = await _client.get(
        Uri.parse('$baseUrl/$resourceType/$id'),
        headers: {
          'Authorization': 'Bearer $accessToken',
          'Accept': 'application/fhir+json',
        },
      );

      if (response.statusCode == 200) {
        return jsonDecode(response.body) as Map<String, dynamic>;
      } else if (response.statusCode == 404) {
        return null;
      } else {
        throw FhirException(
          'Error fetching $resourceType/$id',
          response.statusCode,
          response.body,
        );
      }
    } catch (e) {
      if (e is FhirException) rethrow;
      throw FhirException('Network error fetching resource: $e', 0, null);
    }
  }

  void dispose() {
    _client.close();
  }
}

class FhirException implements Exception {
  final String message;
  final int? statusCode;
  final String? body;

  const FhirException(this.message, this.statusCode, this.body);

  @override
  String toString() =>
      'FhirException: $message (status: $statusCode)';
}
